# 4.6.2. Merging Data - Task

## This script contain following points:

### 1. Import the orders_products_combined dataframe from the pickle file

### 2. Check the shape of the imported dataframe

### 3. Determine a suitable way to combine the orders_products_combined dataframe with your products data set.

### 4. Confirm the results of the merge using the merge flag

### 5. Export the newly created dataframe as ords_prods_merge in a suitable format

## 01. Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

## 02. Importing data sets

In [2]:
# Create path
path = r'C:\Users\Quinn\OneDrive\Documents\CF - Data Analyst\Data Immersion\Instacart Basket Analysis'

In [4]:
# Import orders_products_combined dataset
df_ords_prods_combined = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

In [7]:
# Import checked products dataset
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index_col = 0)

## 03. Check the shape of the imported dataframe

In [9]:
#Check the output of ords_prods_combined
df_ords_prods_combined.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_last_order  first_order  product_id  add_to_cart_order  \
0                    NaN  First Order         196                  1   
1                    NaN  First Order       14084                  2   
2                    NaN  First Order       12427                  3   
3                    NaN  First Order       26088                  4   
4                    NaN  First Order       26405                  5   

   reordered _merge  
0          0   both  
1          0   both  
2          0   both  
3          0   both  
4          0   both

In [10]:
df_ords_prods_combined.shape

(32434489, 11)

In [11]:
# Check the output of checked prods
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [12]:
df_prods.shape

(49672, 5)

## 04. Determine a suitable way to combine the orders_products_combined dataframe with the products data set

### Despite their different shape, both datasets have the shared column "product_id". This should be a fully matching column; hence, we don't need to specify a type of join and inner join will be used in default. The resulting data set will only contain observations included in both input data sets

In [13]:
# Dropping the _merge column df_ords_prods_combine to prevent duplicating columns
df_ords_prods_combined = df_ords_prods_combined.drop(['_merge'], axis = 'columns')

In [16]:
df_ords_prods_combined.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_last_order  first_order  product_id  add_to_cart_order  \
0                    NaN  First Order         196                  1   
1                    NaN  First Order       14084                  2   
2                    NaN  First Order       12427                  3   
3                    NaN  First Order       26088                  4   
4                    NaN  First Order       26405                  5   

   reordered  
0          0  
1          0  
2          0  
3          0  
4          0

In [17]:
# Merge both datasets
df_ords_prods_merge = df_ords_prods_combined.merge(df_prods, on = 'product_id', indicator = True)

In [18]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order      first_order  product_id  add_to_cart_order  \
0                    NaN      First Order         196                  1   
1                   15.0  Repeat Customer         196                  1   
2                   21.0  Repeat Customer         196                  1   
3                   29.0  Repeat Customer         196                  1   
4                   28.0  Repeat Customer         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  
0          0         Soda        77              7     9.0   both  
1          1         Soda        77              7     9.0   both  
2          1         Soda        77              7     9.0   both  
3          1         Soda        77              7     9.0   both  
4          1         Soda        77              7     9.0   both

In [19]:
df_ords_prods_merge.shape

(32404859, 15)

## 05. Confirm the results of the merge using the merge flag

In [20]:
df_ords_prods_merge['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

### The new dataframe has 32,404,859 observations. As mentioned, the result data set only shows observations included in both data set.

## 06. Export the newly created dataframe as ords_prods_merge in a suitable format

### I chose pickle format file because of its advantages in saving dataframe exactly and quickly exporting

In [21]:
# Export merged dataframe into pickle file
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged.pkl'))